### PR TITLE
Set the number of contexts during predict in FiD

### PR DESCRIPTION
**Pull request by mihaela-bornea**
# Pull Request      

## What does this PR do?

Fix for the GenerativeFiDReader.  Set the  number of contexts in the FiD reader based on the number of retrieved passages. 

## Description

Added a setter in the ELI5 FiD preprocessor to allow setting the _max_contexts. The FiD Generative reader can run with different number of contexts compared to training set and it has been tested with less or more number of contexts compared to train set.

```(description)```     

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

**Review by seirasto — APPROVED** [review]
LGTM

**Review by bhavani105 — APPROVED** [review]
LGTM

File: primeqa/components/reader/generative.py [hunk 1]
```diff
@@ -204,7 +204,7 @@ def predict(
             )
         )
 
-        self._preprocessor.max_contexts = len(processed_context)
+        self._preprocessor._max_contexts = len(processed_context)
```
Comment (kpfadnis) [inline primeqa/components/reader/generative.py:207]: @mihaela-bornea I recommend adding a setter function to set the value of `private` member like `_max_contexts`. 

PEP8 naming convention implies that any variable with leading underscore should be considered private and preferably not directly accessed outside.

**Review by kpfadnis — CHANGES_REQUESTED** [review]
I requested a minor change in terms of adding a setter function. Please let me know if you have reservations/concerns about that approach. 

Thanks

File: primeqa/components/reader/generative.py (hunk 1 above)
Comment (mihaela-bornea) [reply primeqa/components/reader/generative.py:207]: Done !

**Review by kpfadnis — APPROVED** [review]
LGTM.

@mihaela-bornea  Can you just updated the PR description or add an comment stating that FiD Generative reader can run with different number of contexts compared to training  set and it has been tested with less or more number of contexts compared to train set.
